src/subject_name/verify.rs

```diff
@@ -30,24 +30,29 @@ pub(crate) fn verify_cert_dns_name(
 ) -> Result<(), Error> {
     let cert = cert.inner();
     let dns_name = untrusted::Input::from(dns_name.as_ref().as_bytes());
-    iterate_names(
+    NameIterator::new(
         Some(cert.subject),
         cert.subject_alt_name,
         SubjectCommonNameContents::Ignore,
-        Err(Error::CertNotValidForName),
-        &mut |name| {
-            let presented_id = match name {
-                GeneralName::DnsName(presented) => presented,
-                _ => return None,
-            };
-
-            match dns_name::presented_id_matches_reference_id(presented_id, dns_name) {
-                Ok(true) => Some(Ok(())),
-                Ok(false) | Err(Error::MalformedDnsIdentifier) => None,
-                Err(e) => Some(Err(e)),
-            }
-        },
     )
+    .find_map(|result| {
+        let name = match result {
```

Comment:
One downside of the stricter iterator chaining is that each of the `NameIterator` users have to deal with error values separately (rather than `iterate_names()` propagating them in one place). However, since the only thing they have to do is yield them to the `find_map()` iterator, that doesn't seem like a big issue.